### ISSUE TITLE
Support (or document non-support) for Fragments

### ISSUE THREAD

**Issue by alexkrolick** [developit/preact#946]
https://github.com/reactjs/reactjs.org/pull/345

React is adding a top-level React.Fragment export that transpiles to JSX as:

```jsx
<>
  <li>Stuff</li>
  <li>Things</li>
</>

React.createElement(React.Fragment, {}, ...)
```

Would Preact be able to support fragments or provide some type of fallback?

**Comment by leeoniya** [comment]
https://github.com/developit/preact/issues/939

**Comment by developit** [comment]
+ #929. I might be working on something.

**Comment by koutsenko** [comment]
[They introduces](https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html) <Fragment> component also. In some cases (e.g. old eslint rules) its more preferrable...

**Comment by developit** [comment]
If you want to use Fragment faked out for now, just do this:

```js
React.Fragment = 'x-fragment';
```

**Comment by CaptainN** [comment]
Where would one put that? `React.Fragment = 'x-fragment';` - somewhere in a webpack config?

**Comment by kurtextrem** [comment]
@CaptainN I guess in index.js or something similar.

**Comment by adamduncan** [comment]
Is lack of support for `Fragment` why we see `<undefined>` elements in output?

<img width="236" alt="screen shot 2018-02-05 at 19 08 00" src="https://user-images.githubusercontent.com/2671053/35824041-739ca8d6-0aa9-11e8-85b1-5c3c8d6a1c60.png">

Would reassigning `React.Fragment` to `'x-fragment'` be needed in the component that makes use of `Fragment`, and still result in extraneous custom elements in the DOM? I.e.

```
import React, { Fragment } from 'react'
React.Fragment = 'x-fragment'
```

**Comment by arusanov** [comment]
@developit any estimate on this feature?

**Comment by phun-ky** [comment]
What is the status of this @developit  ?

**Comment by developit** [comment]
Right now we just have the [hack I posted](https://github.com/developit/preact/issues/946#issuecomment-353151850). Fragment support requires a complete re-architecture of Preact, which I've been working on.  I will be making that work public soon.

**Comment by marvinhagemeister** [comment]
@developit Any to help out with that? I'd love to get involved 🎉

**Comment by developit** [comment]
@marvinhagemeister I just added you as a contributor here. Ping me on slack!

**Comment by marvinhagemeister** [comment]
Just got contributor access and went through a few minor PRs to get started. Implementing `Fragments` shouldn't be that hard. I did a preview implementation for my preact-server renderer here https://github.com/marvinhagemeister/preact-server-renderer/pull/14/files . Basically a `Fragment` is a plain component which is treated differently at the render phase.

Before we can tackle it there are a few tasks, which will make adding them much easier:

1. get dependencies up to date
2. update to `babel 7` because of `<>foo</>` jsx tags. (Fragments aren't much fun without them)
3. actually implement support for fragments.

I'm currently working on the first to and hopefully we can tackle point 3 soon 🎉

**Comment by phun-ky** [comment]
@marvinhagemeister nice work! but I personally feel that we could skip 2. for this fix if possible. I really hope that this fix could be in place before babel@7 enters the playing field, since it's still in beta

**Comment by EmielM** [comment]
I concur with @phun-ky. In my projects we're not using React.Fragment often, but only sporadically where it aids in getting a simpler (flex) layout (ie. preventing extra wrapping). In those few instances, using `<React.Fragment>` is preferred as it makes more explicit what's going on.

**Comment by adamduncan** [comment]
Nice one @marvinhagemeister.  +1 on bypassing point 2. It's not a showstopper after a bit of destructuring.

```
import React, { Fragment } from 'react'

<Fragment>
  ...
</Fragment>
```

**Comment by georgiemathews** [comment]
@marvinhagemeister what’s the progress on this? Is there any way I can help out?

**Comment by marvinhagemeister** [comment]
@georgiemathews We do have Fragments working in a private branch, but we haven't had the time to make it work with hydration yet.

**Comment by prakashsanker** [comment]
@marvinhagemeister  pretty excited for this - when do you think that this will merge?

**Comment by marvinhagemeister** [comment]
@prakashsanker Progress can be tracked here: #1080 🎉

**Comment by andrewkesper** [comment]
I'm able to achieve the desired effect by returning an array of elements. 

Here's a simple example of an HTML definition list:

```jsx
export default class Glossary extends Component {
  state = { 
    glossary: [
      {term: 'HTML', description: 'Hypertext Markup Language'},
      {term: 'JS', description: 'JavaScript'},
      {term: 'CSS', description: 'Cascading Style Sheet'}
    ]
  };
  render({ }, { glossary }) {
    return (
      <dl>
        { glossary.map( g => (
          [ <dt>{g.term}</dt>, <dd>{g.description}</dd> ] // Here is the magic
        )) }
      </dl>
    );
  }
}
```

HTML output:
```html
<dl>

  <dt>HTML</dt>
  <dd>Hypertext Markup Language</dd>

  <dt>JS</dt>
  <dd>JavaScript</dd>

  <dt>CSS</dt>
  <dd>Cascading Style Sheet</dd>

</dl>
```

[**Working demo here**](https://preactjs.com/repl?code=export%20default%20class%20Glossary%20extends%20Component%20%7B%0A%20%20state%20%3D%20%7B%20%0A%20%20%20%20glossary%3A%20%5B%0A%20%20%20%20%20%20%7Bterm%3A%20%27HTML%27%2C%20description%3A%20%27Hypertext%20Markup%20Language%27%7D%2C%0A%20%20%20%20%20%20%7Bterm%3A%20%27JS%27%2C%20description%3A%20%27JavaScript%27%7D%2C%0A%20%20%20%20%20%20%7Bterm%3A%20%27CSS%27%2C%20description%3A%20%27Cascading%20Style%20Sheet%27%7D%0A%20%20%20%20%5D%0A%20%20%7D%3B%0A%20%20render(%7B%20%7D%2C%20%7B%20glossary%20%7D)%20%7B%0A%20%20%20%20return%20(%0A%20%20%20%20%20%20%3Cdl%3E%0A%20%20%20%20%20%20%20%20%7B%20glossary.map(%20g%20%3D%3E%20(%0A%20%20%20%20%20%20%20%20%20%20%5B%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cdt%3E%7Bg.term%7D%3C%2Fdt%3E%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%3Cdd%3E%7Bg.description%7D%3C%2Fdd%3E%0A%20%20%20%20%20%20%20%20%20%20%5D%0A%20%20%20%20%20%20%20%20))%20%7D%0A%20%20%20%20%20%20%3C%2Fdl%3E%0A%20%20%20%20)%3B%0A%20%20%7D%0A%7D)

**Comment by apfelbox** [comment]
@andrewkesper that works if you are **inside** a JSX.Element. What fragments explicitly allow is something else: returning "an array" directly from `render()`, so that you don't need to have a single root element on components (at least non that ends up in the DOM).

**Comment by qodesmith** [comment]
Could we make a bootleg Fragment component like this:

```
import React from 'react'

const Fragment = ({ children }) => React.Children.toArray(children)

export default Fragment
```

The possible downside is that it will auto-assign the key prop to each element. But you essentially get what Fragment does - no containing element. Any thoughts?

**Comment by marvinhagemeister** [comment]
No, current preact expects a DOM node per vnode. This is not true for Fragments where the whole point of them is to not render anything into the DOM. So this is not easily doable (if at all) with the current architecture.

That said we have merged the PR for Fragments into our development repo last week. Fragments will be part of the next major release🎉

It is not available to the public yet and we don't have a release estimate for now. It's progressing nicely and we are currently working through an ever shrinking number of failing test cases and ensuring that all known bugs in the next version are fixed.

**Comment by Ayc0** [comment]
@marvinhagemeister do you know when the new version of Preact with Fragments will be released?

**Comment by calebeby** [comment]
https://twitter.com/marvinhagemeist/status/1056229900158451712

**Comment by andidevi** [comment]
This workaround is closer to React.Fragment and will keep your project compatible with React:
```
import React from 'react';
import pure from 'recompose/pure';
const Fragment = React.Fragment || pure(({children}) => children);
```

**Comment by andidevi** [comment]
There's a bug in preact witch will prevent the rendering of the remaining parts of an children-arrays, if they contain a null value.

**Comment by marvinhagemeister** [comment]
@andidevi Do you have a simple test-case or snippet that we can use to reproduce the issue you are experiencing?

**Comment by IngwiePhoenix** [comment]
So... What's the current status on this? I am currently looking at which library to choose - and I'd rather choose Preact, due to it's size.

**Comment by marvinhagemeister** [comment]
We're getting close to a release, but as with everything the last few percents are the hardest. We basically have 1 bug left and need to get `compat` ready.

**Comment by qodesmith** [comment]
Very excited about this :)

**Comment by kontrollanten** [comment]
I've created a small babel plugin to add support for React.Fragment. What it does is simply to remove the React.Fragment element. I've got it working with `material-ui/icons`. I'll upload it to npm if it's wanted, but now you can use by reference to GH. I've written a readme at https://github.com/kontrollanten/babel-plugin-preact-fragment

**Comment by reznord** [comment]
Fragments are implemented in #1302 (will be available in an alpha release pretty soon)

**Comment by andyfurniss4** [comment]
Any idea when this will be out of alpha?

**Comment by marvinhagemeister** [comment]
@andyfurniss4 At this point we can't give a specific timeframe on when we'll be out of alpha. We'll likely move to beta when `Fragments` are rock solid 👍

**Comment by jessicabyrne** [comment]
We have a project written in Preact that is referencing a component from another library that does not render at all on first render (it will appear if I force a rerender)

```
const TheComponent = React.forwardRef((props, ref) => {
  const { ...rest } = props;

  return (
    <div {..rest} />
});

export default TheComponent;
```

Is this a known issue with Preact X (which we are now on! 🎉 )
What is the most reliable work-around for this right now? I've tried adding `React.Fragment = 'x-fragment'` at the top of the referencing file, but it didn't change anything.

**Comment by marvinhagemeister** [comment]
@jessicabyrne The issue you are describing may not be related at all to `Fragments`. Can you open a new one issue where we can get to the bottom of it together? Maybe we can even reproduce the issue in a [codesandbox](https://codesandbox.io).

**Comment by jessicabyrne** [comment]
@marvinhagemeister I'm convinced it's to do with `Fragments` after doing some more investigating. If I change the referenced code to use a `div` instead, the component renders fine. I did open a new issue though, we can move the conversation there :)
